cfpb/cfgov-refresh
# Adding Babel to frontend workflow; Upgrading Webpack to version 2.

Adding babel to frontend workflow; Upgrading Webpack to version 2.

## Additions

- Added `.babelrc` as a babel Configuration file. Using the `es2015` and  `stage-2` presets. Presets are just sets of transformations. 

## Changes

- Modified `config/webpack-config.js` to do the following:
   - Use the Babel loader
   - Enable the cache property for performance purposes. 
   - Delete the second chunk entry which seems no longer necessary in Webpack 2. 


 - Modified `gulp/tasks/scripts.js` to pass `webpack-stream` a Webpack param. 


 - Updated `package.json` to do the following:
   - Add the necessary modules to support Babel.
   - Remove `slick-carousel` as it's no longer needed.
   - Move `snyk` as it's not necessary to run on Jenkins.

## Testing

- Run `rm -rf node_modules`
- Run `frontend.sh`.
- Browse the site as expected.
- Run `rm -rf node_modules`
- Run `frontend.sh production`.
- Browse the site as expected.

## Screenshots
Before:

<img width="1134" alt="screen shot 2017-02-08 at 12 26 39 pm" src="https://cloud.githubusercontent.com/assets/1696212/22748990/e4cc98ae-edf9-11e6-9e2d-dd79699832b6.png">

After:

<img width="1140" alt="screen shot 2017-02-08 at 11 48 54 am" src="https://cloud.githubusercontent.com/assets/1696212/22748012/b3900fda-edf6-11e6-8e15-4e63368ced59.png">

## Notes

- The key performance bottleneck is the `webpack.optimize.UglifyJsPlugin`.  There are alternatives which uglify in [parallel](https://www.npmjs.com/package/webpack-uglify-parallel), we should investigate if they are worth using.  
- This is an interesting article: https://github.com/lcxfs1991/blog/issues/15.

## Todos

- Converting some modules over to ES6.

## Checklist
